File: leetcode/next_permutation/NextPermutation.cpp
#include <iostream> 
#include <vector> 
#include <algorithm> 

using namespace std; 

class Solution {
public:
    void swap(int &a, int &b) {
        int temp = a; 
        a = b; 
        b = temp; 
    }
    
    void nextPermutation(vector<int> &num) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (num.size() <= 1) {
            return; 
        }
        
        int i = num.size() - 2; 
        while (i >= 0 && (num[i] >= num[i + 1])) {
            i--; 
        }
        
        if (i < 0) {
            reverse(num.begin(), num.end()); 
            return; 
        }
        
        int j = i + 1; 
        while (j < num.size() && num[j] > num[i]) {
            j++; 
        }
        
        swap(num[i], num[j - 1]); 
        reverse(num.begin() + i + 1, num.end()); 
    }
};

int main() {
    Solution sln; 
    vector<int> num; 
    num.push_back(1); 
    num.push_back(5); 
    num.push_back(1);
    sln.nextPermutation(num); 
    for (int i = 0; i < num.size(); i++) {
        cout << num[i] << " "; 
    }
    return 0; 
}

File: leetcode4/NextPermutation.java
public class Solution {
    public void nextPermutation(int[] num) {
        int len = num.length; 
        
        int i = len - 1; 
        
        while (i > 0 && num[i] <= num[i - 1]) {
            i--; 
        }
        
        if (i <= 0) {
            reverse(num, 0, len - 1); 
            return; 
        }
        
        int j = len - 1; 
        
        while (j >= i && num[j] <= num[i - 1]) {
            j--; 
        }
        
        int temp = num[i - 1]; 
        num[i - 1] = num[j]; 
        num[j] = temp; 
        
        reverse(num, i, len - 1); 
    }
    
    void reverse(int[] num, int i, int j) {
        while (i < j) {
            int temp = num[i]; 
            num[i] = num[j]; 
            num[j] = temp; 
            i++; 
            j--; 
        }
    }
}
